common/types.ts

```diff
@@ -0,0 +1,15 @@
+/**
+ * A mapping between the library name and the target address location.
+ *
+ * Containing support for two level configuration.
+ */
+export interface LibraryPlaceholdersMapping {
+  [libraryName: string]: string | { [location: string]: string };
```

Comment:
The type is to support both methods which are currently passed in.

`{"lib.sol": {"L": "0x..."}}`
and 
`{"lib.sol": "0x..."}`

which I experienced and gained from writing the changes the tests. If i'm wrong we can correct the names.

Reply:
```
// kind 1: {"lib.sol:L": "0x..."}
// kind 2: {"lib.sol": {"L": "0x..."}}

// we can specify the requirement for `L` if its always going to be `L` or change it to `[x: string]`
type: { [x: string]: string } | { [x: string]: { "L": string }  } 

```

Reply:
Unfortunately the current code supports mixing the two. E.g. this is valid input:
```json
{
    "lib.sol:L1": "0x...",
    "lib.sol:L2": "0x...",
    "lib.sol": {"L3": "0x..."}
}
```
On the other hand, we don't even support the first kind in Standard JSON and I doubt any compiler version supported both so we might be fine changing it. It's technically a breaking change for solc-js though. @axic what do you think?

So yeah, what you did in the PR is correct, just names are a bit misleading. We could do this but it's a bit ugly:
```typescript
export interface LibraryAddresses {
  [qualifiedNameOrSourceUnit: string]: string | { [unqualifiedLibraryName: string]: string };
```
On the other hand the ugliness at least does let you know that there's something wonky going on here and you should watch your step :)

The only other idea that comes to mind is to give the first level some generic name not to suggest it's necessarily one or the other:
```typescript
export interface LibraryAddresses {
  [firstLevelID: string]: string | { [unqualifiedLibraryName: string]: string };
```

Reply:
I'm not sure you understand, "Unfortunately the current code supports mixing the two. E.g. this is valid input:" this type supports both. 

```
{
    "lib.sol:L1": "0x...",
    "lib.sol:L2": "0x...",
    "lib.sol": {"L3": "0x..."}
}
```

Its covered. And its not enforcing you to do one or the other since the internals handles the differences. Too much overthinking of type here imo.

Reply:
Can you explain how it supports both?
```
type: { [x: string]: string } | { [x: string]: { [x: string]: string }  } 
```
Doesn't the `|` here mean that it must have one signature or the other and not that you can mix and match fields allowed in both?

In any case, unless it supports both, that suggestion does not match what the library supports.

The problem still stands, names and the docstring here are ambiguous and need to be adjusted.

Reply:
Since the `libraries: LibraryAddresses` is an object, it's stating that the content of each item within that object can be one or the other. A single item cannot be both so this is fine. `In any case, unless it supports both, that suggestion does not match what the library supports.` - you will have to be more clean on this since I have made all the required changes

Reply:
> Since the `libraries: LibraryAddresses` is an object, it's stating that the content of each item within that object can be one or the other. A single item cannot be both so this is fine.

I think we might be talking about different things here. I get the feeling you're commenting on the form you used in the PR and not on the one I cited. But no matter, that's not all that important so let's drop it. Perhaps I'm just misunderstanding this. I was just giving feedback on your suggestion from https://github.com/ethereum/solc-js/pull/604#discussion_r807076160 because you asked for it.

The version in the PR is fine. I'd just replace the `two level configuration` part in the docstring with the example of two kinds you have above.

Reply:
sounds good,

i'll add this example

```
{
    "lib.sol:L1": "0x...",
    "lib.sol:L2": "0x...",
    "lib.sol": {"L3": "0x..."}
}
```

Reply:
done